### PR TITLE
Part 1 : message_topic section_uuid

### PR DESCRIPTION
**Pull request by sorja**
Example output for FRA 2020 message_topic after migration step:

| id | country\_iso | key | status | type | section\_uuid |
| :--- | :--- | :--- | :--- | :--- | :--- |
| 2093 | VNM | 718-class-02d4bcc9-d7b1-4ce3-8ad0-42619ddc890d-extentOfForest | resolved | review | 73e335e2-8079-4438-8644-0fdab3ea30cf |
| 2094 | VNM | 719-class-01a8b58b-b801-48ad-bbc2-fb9d6f7e6744-extentOfForest | opened | review | 73e335e2-8079-4438-8644-0fdab3ea30cf |
| 2095 | VNM | 720-class-c40d32c2-7e9c-4983-95a6-5ab5e2613a05-extentOfForest | resolved | review | 73e335e2-8079-4438-8644-0fdab3ea30cf |
| 2096 | VNM | 722-class-f15e3125-0d6c-4949-b2a2-ff3a38bef5df-extentOfForest | resolved | review | 73e335e2-8079-4438-8644-0fdab3ea30cf |
| 2097 | VNM | 721-class-a91a33cc-147f-43f6-bc6c-728b6225a2d5-extentOfForest | resolved | review | 73e335e2-8079-4438-8644-0fdab3ea30cf |
| 2066 | VNM | 718-class-02d4bcc9-d7b1-4ce3-8ad0-42619ddc890d-forestCharacteristics | resolved | review | 0ec33ee4-3cd5-4261-a1dd-16afa1783381 |

where uuid:

| uuid | section\_name |
| :--- | :--- |
| 73e335e2-8079-4438-8644-0fdab3ea30cf | extentOfForest |
| 0ec33ee4-3cd5-4261-a1dd-16afa1783381 | forestCharacteristics |

### CODE REVIEW

File: src/test/migrations/steps/20230315113503-step-message_topic-section_uuid.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import { BaseProtocol, DB, Schemas } from '@server/db'
+
+export default async (client: BaseProtocol) => {
+  const updateTargets = await client.map(
+    `select a.props ->> 'name' as assessment_name, c.name as cycle_name
+     from public.assessment a
+              left join public.assessment_cycle c on a.id = c.assessment_id;`,
+    [],
+    (row) => Objects.camelize(row)
+  )
```
Comment (minotogna) [inline src/test/migrations/steps/20230315113503-step-message_topic-section_uuid.ts]: would it make sense to add a ```AssessmentController.getAll()=>Array<Assessment> ```?

File: src/test/migrations/steps/20230315113503-step-message_topic-section_uuid.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      from assessment_fra.row r
+               left join ${schemaName}.table t on r.table_id = t.id
+               left join ${schemaName}.table_section ts on t.table_section_id = ts.id
+               left join ${schemaName}.section s on ts.section_id = s.id
+      where r.uuid::text in (select key from ${schemaCycle}.message_topic)
+        and mt.key = r.uuid::text;
+  `)
+  }
+
+  // Only relevant for assessment_fra_2025
```
Comment (minotogna) [inline src/test/migrations/steps/20230315113503-step-message_topic-section_uuid.ts]: I think we should think about all cycles directly.

**Review by minotogna — COMMENTED** [review]
the tests do not pass...

File: src/test/migrations/steps/20230315113503-step-message_topic-section_uuid.ts (hunk 2 above)
Comment (minotogna) [reply src/test/migrations/steps/20230315113503-step-message_topic-section_uuid.ts]: do commentable-description exists only in 2025?

Comment (sorja) [reply src/test/migrations/steps/20230315113503-step-message_topic-section_uuid.ts]: seems so, assessment_fra_2020 after running yarn migrate-fra and migration-steps:

image

File: src/test/migrations/steps/20230315113503-step-message_topic-section_uuid.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
+// import { Objects } from '@utils/objects'
+
+import { AssessmentController } from '@server/controller/assessment'
+import { BaseProtocol, DB, Schemas } from '@server/db'
+
+export default async (client: BaseProtocol) => {
+  const assessments = await AssessmentController.getAll({ metaCache: false }, client)
+  const schemas = assessments.flatMap((assessment) => {
```
Comment (sorja) [inline src/test/migrations/steps/20230315113503-step-message_topic-section_uuid.ts:6]: `const schemas =`
```
      [
        {
          schemaCycle: 'assessment_panEuropean_2025',
          schemaName: 'assessment_panEuropean'
        },
        {
          schemaCycle: 'assessment_panEuropean_2020',
          schemaName: 'assessment_panEuropean'
        },
        { schemaCycle: 'assessment_fra_2025', schemaName: 'assessment_fra' },
        { schemaCycle: 'assessment_fra_2020', schemaName: 'assessment_fra' }
      ]
```

File: src/server/repository/assessment/assessment/readAll.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+import { Assessment } from '@meta/assessment'
+
+import { BaseProtocol, DB } from '@server/db'
+
+const fields: Array<string> = ['id', 'uuid', 'props']
+
+const selectFields = fields.map((f) => `a.${f}`).join(',')
+
+export const readAll = async (
```
Comment (minotogna) [inline src/server/repository/assessment/assessment/readAll.ts]: I would call this getAll

File: src/server/repository/assessment/assessment/readAll.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import { Objects } from '@utils/objects'
+
+import { Assessment } from '@meta/assessment'
+
+import { BaseProtocol, DB } from '@server/db'
+
+const fields: Array<string> = ['id', 'uuid', 'props']
```
Comment (minotogna) [inline src/server/repository/assessment/assessment/readAll.ts]: Can't this be shared with getOne ?

File: src/server/repository/assessment/assessment/readAll.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    `
+        select ${selectFields},
+               jsonb_agg(to_jsonb(ac.*)) as cycles
+            ${props.metaCache ? `, meta_cache as metacache` : ''}
+        from assessment a
+                 left join assessment_cycle ac on a.id = ac.assessment_id
+        group by ${selectFields};
+    `,
+    [],
+    ({ metacache, ...assessment }) => {
```
Comment (minotogna) [inline src/server/repository/assessment/assessment/readAll.ts]: Can we have an assessment adapter ?

Comment (sorja) [reply src/server/repository/assessment/assessment/readAll.ts]: We already had!

File: src/server/controller/assessment/getOne.ts [hunk 7]
```diff
@@ -28,7 +28,7 @@ export const getOneWithCycle = async (
   client: BaseProtocol = DB
 ): Promise<{ assessment: Assessment; cycle: Cycle }> => {
   const { cycleName, cycleUuid } = props
-  const assessment = await AssessmentRepository.read(props, client)
+  const assessment = await AssessmentRepository.get(props, client)
```
Comment (minotogna) [inline src/server/controller/assessment/getOne.ts]: I would call this getOne

File: src/test/migrations/steps/20230315113503-step-message_topic-section_uuid.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+                         from ${schemaCycle}.descriptions d
+                         where d.value -> 'dataSources' is not null)
+    update
+        ${schemaCycle}.message_topic mt
+    set
+        section_uuid = s.uuid
+    from datasources left join ${schemaName}.section s on section_name = s.props ->> 'name'
+        where datasources.uuid = mt.key;
+        `)
+  }
```
Comment (minotogna) [inline src/test/migrations/steps/20230315113503-step-message_topic-section_uuid.ts:87]: is other migration still missing, right? would it make sense just 1 migration step?

File: src/server/controller/assessment/getOne.ts (hunk 7 above)
Comment (sorja) [reply src/server/controller/assessment/getOne.ts]: good catch, typo

File: src/test/migrations/steps/20230315113503-step-message_topic-section_uuid.ts (hunk 8 above)
Comment (sorja) [reply src/test/migrations/steps/20230315113503-step-message_topic-section_uuid.ts:87]: Right! Added

**Review by rfontanarosa — APPROVED** [review]

**Pull request merged**